docs/performance.rst

```diff
@@ -0,0 +1,139 @@
+===============================
+LASER Performance Optimization
+===============================
+
+It's fairly easy to generate working LASER model code that uses NumPy and performs well for populations of 100,000 agents. We expect LASER code to perform efficiently for up to **200,000,000 agents**. The challenge is how to get there.
+
+Identifying Performance Bottlenecks
+-----------------------------------
+
+Typically, we do **not** recommend running the code through a profiler—at least not initially. Instead, we take advantage of **LASER's highly modular structure** and **AI-driven optimization**.
+
+The first step is to **add simple timing code** that tracks the total time spent in each component over a realistic simulation. Then, **plot a pie chart** at the end to visualize where the most time is spent. A simple way to track execution time is using the ``time`` package. Example code:
+
+.. code-block:: text
+
+    def run(self):
+        self.component_times = {component.__class__.__name__: 0 for component in self.components}
+        self.component_times["reporting"] = 0
+        for tick in tqdm(range(self.pars.timesteps)):
+            for component in self.components:
+                start_time = time.time()  # Start timing for the component
+                component.step()
+                end_time = time.time()  # End timing for the component
+
+                # Accumulate the time taken for this component
+                elapsed_time = end_time - start_time
+                component_name = component.__class__.__name__
+
+This often reveals **the top 1 to 3 performance bottlenecks**. Focus first on the biggest offender—it provides the most opportunity for speedup. Often, the largest bottleneck is **not** what you might instinctively expect. Avoid optimizing a component only to find out it contributes just **5% of the total runtime**. Also, make sure that your reporting code is being measured and reported, ideally in its own 'bucket'. This may be easier or harder depending on how you are doing reporting. Since reporting usually involves counting over the entire population, it usually shows up as a hotspot sooner or later. Fortunately, it's usually fairly easy to speed up. Or even eliminate.
```

Comment:
<strike>**5% of the total runtime**</strike> **a small percentage of the total runtime**. A modest improvement in the runtime of an "expensive" component is often more effective than spending a lot of time on highly optimizing a component which only accounts for a small fraction of runtime.